File: src/main/java/dev/psygamer/wireframe/util/collection/FreezableMap.java
package dev.psygamer.wireframe.util.collection;

import dev.psygamer.wireframe.util.helper.ICloneable;
import dev.psygamer.wireframe.util.helper.IFreezable;

import com.google.common.collect.ImmutableMap;

import java.util.Set;
import java.util.Map;
import java.util.Collection;

public abstract class FreezableMap <K, V> implements Map<K, V>, IFreezable, ICloneable<FreezableMap<K, V>> {
	
	private final Map<K, V> map;
	private volatile boolean frozen = false;
	
	protected FreezableMap(final Map<K, V> map) {
		this.map = map;
	}
	
	@Override
	public int size() {
		return this.map.size();
	}
	
	@Override
	public boolean isEmpty() {
		return this.map.isEmpty();
	}
	
	@Override
	public boolean containsKey(final Object key) {
		return this.map.containsKey(key);
	}
	
	@Override
	public boolean containsValue(final Object value) {
		return this.map.containsValue(value);
	}
	
	@Override
	public V get(final Object key) {
		return this.map.get(key);
	}
	
	@Override
	public V put(final K key, final V value) {
		IFreezable.throwIfFrozen(this);
		return this.map.put(key, value);
	}
	
	@Override
	public V remove(final Object key) {
		IFreezable.throwIfFrozen(this);
		return this.map.remove(key);
	}
	
	@Override
	public void putAll(final Map<? extends K, ? extends V> m) {
		IFreezable.throwIfFrozen(this);
		this.map.putAll(m);
	}
	
	@Override
	public void clear() {
		IFreezable.throwIfFrozen(this);
		this.map.clear();
	}
	
	@Override
	public Set<K> keySet() {
		return this.map.keySet();
	}
	
	@Override
	public Collection<V> values() {
		return this.map.values();
	}
	
	@Override
	public Set<Entry<K, V>> entrySet() {
		return this.map.entrySet();
	}
	
	@Override
	public void freeze() {
		this.frozen = true;
	}
	
	@Override
	public boolean isFrozen() {
		return this.frozen;
	}
	
	public ImmutableMap<K, V> toImmutable() {
		return ImmutableMap.copyOf(this);
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/util/collection/FreezableMap.kt
package dev.psygamer.wireframe.util.collection

import com.google.common.collect.ImmutableMap
import dev.psygamer.wireframe.util.helper.IFreezable

abstract class FreezableMap<K, V> protected constructor(private val map: MutableMap<K, V>) : MutableMap<K, V>, IFreezable {
	
	@Volatile
	final override var frozen = false
		private set
	
	override val size
		get() = map.size
	
	override fun put(key: K, value: V): V? {
		IFreezable.throwIfFrozen(this)
		return map.put(key, value)
	}
	
	override fun putAll(from: Map<out K, V>) {
		IFreezable.throwIfFrozen(this)
		map.putAll(from)
	}
	
	override fun remove(key: K): V? {
		IFreezable.throwIfFrozen(this)
		return map.remove(key)
	}
	
	override fun clear() {
		IFreezable.throwIfFrozen(this)
		map.clear()
	}
	
	override fun isEmpty() =
		map.isEmpty()
	
	override fun containsKey(key: K) =
		map.containsKey(key)
	
	override fun containsValue(value: V) =
		map.containsValue(value)
	
	override operator fun get(key: K) =
		map[key]
	
	override val keys
		get() = map.keys
	
	override val values
		get() = map.values
	
	override val entries
		get() = map.entries
	
	override fun freeze() {
		frozen = true
	}
	
	fun toImmutable(): ImmutableMap<K, V> {
		return ImmutableMap.copyOf(this)
	}
}